### ISSUE TITLE
AccessViolationException in PEHeader.FromBuffer

### ISSUE THREAD

**Issue by discostu105** [Microsoft/clrmd#159]
I'm using CLRMD 1.0 and since recently I'm seeing this exception (not for all crash dumps):

```
ERROR: WinDbgAnalyzer failed: System.AccessViolationException: Attempted to read or write protected memory. This is often an indication that other memory is corrupt.
   at Microsoft.Diagnostics.Runtime.Interop.IDebugControl6.Execute(DEBUG_OUTCTL OutputControl, String Command, DEBUG_EXECUTE Flags)
   at SuperDump.Analyzers.WinDbgAnalyzer.ExecuteCommand(IDebugControl6 debugControl, String command, String comment) in C:\workspaces\priv\superdump-internal.git\src\SuperDump\Analyzers\WinDbgAnalyzer.cs:line 48
   at SuperDump.Analyzers.WinDbgAnalyzer.Analyze(IDebugControl6 debugControl, String logfilepath) in C:\workspaces\priv\superdump-internal.git\src\SuperDump\Analyzers\WinDbgAnalyzer.cs:line 29
   at SuperDump.Analyzers.WinDbgAnalyzer.Analyze() in C:\workspaces\priv\superdump-internal.git\src\SuperDump\Analyzers\WinDbgAnalyzer.cs:line 17
   at SuperDump.Program.<>c.<RunAnalysis>b__7_0() in C:\workspaces\priv\superdump-internal.git\src\SuperDump\Program.cs:line 67
   at SuperDump.Program.RunSafe(String name, Action action) in C:\workspaces\priv\superdump-internal.git\src\SuperDump\Program.cs:line 213
INFO: created runtime with version v2.0.50727.8793
ERROR: WinDbgAnalyzer failed: System.AccessViolationException: Attempted to read or write protected memory. This is often an indication that other memory is corrupt.
   at Microsoft.Diagnostics.Runtime.Utilities.PEHeader.FromBuffer(PEBuffer buffer, Boolean virt)
   at Microsoft.Diagnostics.Runtime.Utilities.PEFile.Init(Stream stream, String filePath, Boolean virt, PEBuffer buffer, PEHeader header)
   at Microsoft.Diagnostics.Runtime.Desktop.DesktopModule.get_Pdb()
   at SuperDump.ModelHelpers.ModelHelpers.ToSDModel(ClrModule module) in C:\workspaces\priv\superdump-internal.git\src\SuperDump\ModelHelpers\ModelHelpers.cs:line 130
   at SuperDump.ModelHelpers.ModelHelpers.ToSDModel(ClrAppDomain domain) in C:\workspaces\priv\superdump-internal.git\src\SuperDump\ModelHelpers\ModelHelpers.cs:line 68
   at SuperDump.Analyzers.SystemAnalyzer.GetSharedDomain() in C:\workspaces\priv\superdump-internal.git\src\SuperDump\Analyzers\SystemAnalyzer.cs:line 55
   at SuperDump.Analyzers.SystemAnalyzer.InitSystemInfo() in C:\workspaces\priv\superdump-internal.git\src\SuperDump\Analyzers\SystemAnalyzer.cs:line 35
   at SuperDump.Analyzers.SystemAnalyzer..ctor(DumpContext context) in C:\workspaces\priv\superdump-internal.git\src\SuperDump\Analyzers\SystemAnalyzer.cs:line 23
   at SuperDump.Program.<>c__DisplayClass7_0.<RunAnalysis>b__1() in C:\workspaces\priv\superdump-internal.git\src\SuperDump\Program.cs:line 79
   at SuperDump.Program.RunSafe(String name, Action action) in C:\workspaces\priv\superdump-internal.git\src\SuperDump\Program.cs:line 213
```

Cannot share the dump file for privacy reasons.

Actually, there are two different stacktraces with AccessViolationException, but both are from the same crash-dump, so they might be related. CLR version of the crashed process = 2.0.50727.8793

Here is the code calling CLRMD: 
 * https://github.com/Dynatrace/superdump/blob/master/src/SuperDump/Analyzers/WinDbgAnalyzer.cs#L29
 * https://github.com/Dynatrace/superdump/blob/master/src/SuperDump/ModelHelpers/ModelHelpers.cs#L130

**Comment by leculver** [comment]
Thanks for the report.  I tried to avoid it, but I think I'm going to have to replace that PEFile implementation.  It was pulled from some old code in a parallel project (PerfView) a while back.  It would be a lot better if we just didn't use unsafe/fixed/weird buffers like that code does and just read from a stream into defined structs...

I will add this to my todo list.   I'm on vacation now but I _should_ have some time at the end of the month for a small sprint of work.

**Comment by discostu105** [comment]
Ok, appreciate it!

Just for completeness, I've found a similar problem also in a V4 runtime.

```
ERROR: WinDbgAnalyzer failed: System.AccessViolationException: Attempted to read or write protected memory. This is often an indication that other memory is corrupt.
   at Microsoft.Diagnostics.Runtime.Interop.IDebugControl6.Execute(DEBUG_OUTCTL OutputControl, String Command, DEBUG_EXECUTE Flags)
   at SuperDump.Analyzers.WinDbgAnalyzer.ExecuteCommand(IDebugControl6 debugControl, String command, String comment) in C:\workspaces\priv\superdump-internal.git\src\SuperDump\Analyzers\WinDbgAnalyzer.cs:line 48
   at SuperDump.Analyzers.WinDbgAnalyzer.Analyze(IDebugControl6 debugControl, String logfilepath) in C:\workspaces\priv\superdump-internal.git\src\SuperDump\Analyzers\WinDbgAnalyzer.cs:line 29
   at SuperDump.Analyzers.WinDbgAnalyzer.Analyze() in C:\workspaces\priv\superdump-internal.git\src\SuperDump\Analyzers\WinDbgAnalyzer.cs:line 17
   at SuperDump.Program.<>c.<RunAnalysis>b__7_0() in C:\workspaces\priv\superdump-internal.git\src\SuperDump\Program.cs:line 67
   at SuperDump.Program.RunSafe(String name, Action action) in C:\workspaces\priv\superdump-internal.git\src\SuperDump\Program.cs:line 213
INFO: created runtime with version v4.0.30319.1026
ERROR: Exception creating CLR Runtime
ERROR: Object reference not set to an instance of an object.
   at Microsoft.Diagnostics.Runtime.Desktop.LegacyRuntime.GetSvrHeapDetails(UInt64 addr)
   at Microsoft.Diagnostics.Runtime.RuntimeBase.GetHeaps(SubHeap[]& heaps)
   at Microsoft.Diagnostics.Runtime.HeapBase.InitSegments(RuntimeBase runtime)
   at Microsoft.Diagnostics.Runtime.Desktop.DesktopGCHeap..ctor(DesktopRuntimeBase runtime)
   at Microsoft.Diagnostics.Runtime.Desktop.DesktopRuntimeBase.CreateHeap()
   at System.Lazy`1.CreateValue()
   at System.Lazy`1.LazyInitValue()
   at System.Lazy`1.get_Value()
   at Microsoft.Diagnostics.Runtime.Desktop.DesktopRuntimeBase.get_Heap()
   at SuperDump.Program.SetupCLRRuntime() in C:\workspaces\priv\superdump-internal.git\src\SuperDump\Program.cs:line 176

--- General System Info ---
Dump was taken at: 12/5/2018 10:18:21 PM UTC +01:00:00
System up-time: Could not be obtained.
OS version: kernel32.dll version: 6.1.7601.24291 (win7sp1_ldr_escrow.181110-1429)
Process architecture (read from dump): X86
Number of logical processing units in target system: 4

--- CLR version list ---
Found CLR Version:v4.0.30319.1026
Filesize:  671000
Timestamp: 53B50961
Dac File:  mscordacwks_X86_X86_4.0.30319.1026.dll

--- END CLR version list ---
```

Although the second exception (`LegacyRuntime.GetSvrHeapDetails`) is actually a different one (separate bug?).

**Comment by leculver** [comment]
@discostu105: I have just merged a change that should resolve the first problem.  I've implemented a new `PEImage` reader that's much safer than the old one.  I've marked `PEFile` as obsolete.  It'll be a while before I remove it from the runtime (ClrMD 1.2 or so), but all uses of the old class have been removed.

The second callstack is tough to say.  It's possible that either this issue or one of the other issues in `LegacyRuntime` that's been fixed could all cause that.  (It looks like native heap corruption to me, just from the callstack.)

If you have a chance I'd appreciate it if you gave the current ClrMD master branch a try, if you are set up for that.  (Sorry, it'll still be a while before we have daily NuGet packages being produced per-checkin.)  I'm curious if you run into any more issues with that change.
